src/server/game/Handlers/ChallengeHandler.cpp

```diff
@@ -0,0 +1,88 @@
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "WorldSession.h"
+#include "ChallengePackets.h"
+#include "Item.h"
+#include "Group.h"
+#include "ObjectAccessor.h"
+#include <GameTime.h>
+#include "AreaTrigger.h"
+#include "GameObject.h"
+
+void WorldSession::HandleStartChallengeMode(WorldPackets::Challenge::StartChallengeMode& startChallengeMode)
+{
+    Player* owner = GetPlayer();
+
+    if (!owner)
+        return;
+
+    InstanceMap* map = owner->GetMap()->ToInstanceMap();
+
+    if (!map)
+        return;
+
+    Item const* keystone = owner->GetItemByPos(startChallengeMode.Bag, startChallengeMode.Slot);
+
+    if (!keystone)
+        return;
+
+    uint32 challengeId = keystone->GetModifier(ITEM_MODIFIER_CHALLENGE_MAP_CHALLENGE_MODE_ID);
+    uint32 challengeLevel = keystone->GetModifier(ITEM_MODIFIER_CHALLENGE_KEYSTONE_LEVEL);
+
+    Group* group = owner->GetGroup();
+
+    auto StartChallengeFor = [challengeId, map](Player* target)
+    {
+        WorldPackets::Challenge::ChangePlayerDifficultyResult changePlayerDiffResultFirst;
+
+        changePlayerDiffResultFirst.Result = WorldPackets::Challenge::ChangePlayerDifficultyResult::LoadingScreenEnable;
+        changePlayerDiffResultFirst.Unused = false;
+        changePlayerDiffResultFirst.NextDifficultyChangeTime = GameTime::GetGameTime() + 2;
+
+        target->GetSession()->SendPacket(changePlayerDiffResultFirst.Write());
+        target->SetDungeonDifficultyID(Difficulty::DIFFICULTY_MYTHIC_KEYSTONE);
+
+        map->RespawnAllObjects();
+
+        WorldPackets::Challenge::ChangePlayerDifficultyResult changePlayerDiffResultSecond;
+
+        changePlayerDiffResultSecond.Result = WorldPackets::Challenge::ChangePlayerDifficultyResult::Sucess;
+        changePlayerDiffResultSecond.MapId = map->GetId();
+        changePlayerDiffResultSecond.DifficultyRecID = Difficulty::DIFFICULTY_MYTHIC_KEYSTONE;
+
+        target->GetSession()->SendPacket(changePlayerDiffResultSecond.Write());
+
+        target->TeleportTo(map->GetId(), target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), target->GetOrientation());
+    };
+
+    if (group)
+    {
+        group->SetDungeonDifficultyID(Difficulty::DIFFICULTY_MYTHIC_KEYSTONE);
+
+        Group::MemberSlotList const& members = group->GetMemberSlots();
+
+        for (auto itr = members.begin(); itr != members.end(); ++itr)
+        {
+            if (Player* p = ObjectAccessor::FindPlayer((*itr).guid))
```

Comment:
it doesn't hurt to name the pointer player, having single letters is painful to read later